packages/api/amplify_api/lib/src/graphql/ws/websocket_message_stream_transformer.dart

```diff
@@ -0,0 +1,54 @@
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:amplify_api/src/util.dart';
+import 'package:amplify_core/amplify_core.dart';
+import 'websocket_message.dart';
+
+import '../graphql_response_decoder.dart';
+
+class WebSocketMessageStreamTransformer
+    extends StreamTransformerBase<dynamic, WebSocketMessage> {
+  const WebSocketMessageStreamTransformer();
+
+  @override
+  Stream<WebSocketMessage> bind(Stream<dynamic> stream) {
+    return stream.cast<String>().map<Map>((str) {
+      return json.decode(str) as Map;
+    }).map(WebSocketMessage.fromJson);
+  }
+}
+
+class WebSocketSubscriptionStreamTransformer<T>
+    extends StreamTransformerBase<WebSocketMessage, GraphQLResponse<T>> {
+  final GraphQLRequest<T> request;
+  final void Function()? onEstablished;
+
+  const WebSocketSubscriptionStreamTransformer(
+      this.request, this.onEstablished);
+
+  @override
+  Stream<GraphQLResponse<T>> bind(Stream<WebSocketMessage> stream) async* {
+    await for (var event in stream) {
+      switch (event.messageType) {
+        case MessageType.startAck:
+          onEstablished?.call();
+          break;
+        case MessageType.data:
+          final payload = event.payload as SubscriptionDataPayload;
+          final errors = deserializeGraphQLResponseErrors(payload.toJson());
+          yield GraphQLResponseDecoder.instance.decode<T>(
+              request: request,
+              data: json.encode(payload.data),
```

Comment:
Ew so this data will have gone through 4 encode/decode passes before finally being returned to the user?

Reply:
yeah, not ideal, still working on improving this.

Reply:
I would rather improve this in a separate PR as this is all based on this method taking a string (based on method channel implementation) and changing it would touch several places. I think my simply making this method take a map (could be with errors as well) it would reduce 2 of those conversions in many cases. Would TODO comment here suffice for that? Do you have other suggestions for simplifying it?

Reply:
Either a TODO or FIXME. Let's not lose track of these changes though. Why are we waiting to remove method channel code?

Reply:
added TODO